File: ProyectoFinalEDD/java/src/HashTable.java
import java.io.*;

class Node {

    Student data;

    Node next;

    public Node(Student data) {

        this.data = data;

        this.next = null;
    }
}

public class HashTable {

    int SIZE = 10;

    Node[] table;

    public HashTable() {

        table = new Node[SIZE];
    }

    public int hashFunction(int id) {

        return id % SIZE;
    }

    public boolean exists(int id) {

        int index =
        hashFunction(id);

        Node temp =
        table[index];

        while(temp != null) {

            if(temp.data.student_id == id) {

                return true;
            }

            temp = temp.next;
        }

        return false;
    }

    public void insert(Student student) {

        int index =
        hashFunction(student.student_id);

        Node newNode =
        new Node(student);

        if(table[index] == null) {

            table[index] = newNode;
        }

        else {

            Node temp =
            table[index];

            while(temp.next != null) {

                temp = temp.next;
            }

            temp.next = newNode;
        }
    }

    public void loadCSV(String filename) {

        try {

            BufferedReader br =
            new BufferedReader(
            new FileReader(filename));

            String line;

            while((line = br.readLine())
                   != null) {

                String[] data =
                line.split(",");

                int id =
                Integer.parseInt(data[0]);

                String nombre =
                data[1];

                String carrera =
                data[2];

                int semestre =
                Integer.parseInt(data[3]);

                float gpa =
                Float.parseFloat(data[4]);

                int skill =
                Integer.parseInt(data[5]);

                Student student =
                new Student(
                id,
                nombre,
                carrera,
                semestre,
                gpa,
                skill);

                insert(student);
            }

            br.close();

            System.out.println(
            "\nCSV cargado correctamente");
        }

        catch(Exception e) {

            System.out.println(
            "\nERROR REAL:");

            e.printStackTrace();
        }
    }

    public void remove(int id) {

        int index =
        hashFunction(id);

        Node temp =
        table[index];

        Node prev = null;

        while(temp != null &&
              temp.data.student_id != id) {

            prev = temp;

            temp = temp.next;
        }

        if(temp == null) {

            System.out.println(
            "\nEstudiante no encontrado");

            return;
        }

        if(prev == null) {

            table[index] =
            temp.next;
        }

        else {

            prev.next =
            temp.next;
        }

        System.out.println(
        "\nEstudiante eliminado correctamente");
    }

    public void update(
        int id,
        String nombre,
        String carrera,
        int semestre,
        float gpa,
        int skill
    ) {

        int index =
        hashFunction(id);

        Node temp =
        table[index];

        while(temp != null) {

            if(temp.data.student_id == id) {

                temp.data.full_name =
                nombre;

                temp.data.career =
                carrera;

                temp.data.semester =
                semestre;

                temp.data.gpa =
                gpa;

                temp.data.skill_score =
                skill;

                System.out.println(
                "\nEstudiante actualizado");

                return;
            }

            temp = temp.next;
        }

        System.out.println(
        "\nEstudiante no encontrado");
    }

    public int countStudents() {

        int count = 0;

        for(int i = 0; i < SIZE; i++) {

            Node temp =
            table[i];

            while(temp != null) {

                count++;

                temp = temp.next;
            }
        }

        return count;
    }

    public float averageGPA() {

        float sum = 0;

        int count = 0;

        for(int i = 0; i < SIZE; i++) {

            Node temp =
            table[i];

            while(temp != null) {

                sum += temp.data.gpa;

                count++;

                temp = temp.next;
            }
        }

        if(count == 0) {

            return 0;
        }

        return sum / count;
    }

    public void display() {

        System.out.println(
        "\n===== HASH TABLE =====");

        for(int i = 0; i < SIZE; i++) {

            System.out.print(
            "\nBucket " + i + ": ");

            Node temp =
            table[i];

            while(temp != null) {

                System.out.print(
                "[" +
                temp.data.student_id +
                " - " +
                temp.data.full_name +
                "] -> ");

                temp = temp.next;
            }

            System.out.print("NULL");
        }

        System.out.println();
    }

    public void search(int id) {

        int index =
        hashFunction(id);

        Node temp =
        table[index];

        while(temp != null) {

            if(temp.data.student_id == id) {

                System.out.println(
                "\n===== ESTUDIANTE =====");

                System.out.println(
                "ID: " +
                temp.data.student_id);

                System.out.println(
                "Nombre: " +
                temp.data.full_name);

                System.out.println(
                "Carrera: " +
                temp.data.career);

                System.out.println(
                "Semestre: " +
                temp.data.semester);

                System.out.println(
                "GPA: " +
                temp.data.gpa);

                System.out.println(
                "Skill Score: " +
                temp.data.skill_score);

                return;
            }

            temp = temp.next;
        }

        System.out.println(
        "\nEstudiante no encontrado");
    }
}

File: ProyectoFinalEDD/source/HashTable.cpp
#include "../headers/HashTable.h"

HashTable::HashTable() {

    collisions = 0;

    for(int i = 0; i < TABLE_SIZE; i++) {

        table[i] = nullptr;
    }
}

int HashTable::hashFunction(int id) {

    return id % TABLE_SIZE;
}

bool HashTable::exists(int id) {

    int index = hashFunction(id);

    Node* temp = table[index];

    while(temp != nullptr) {

        if(temp->data.student_id == id) {

            return true;
        }

        temp = temp->next;
    }

    return false;
}

void HashTable::insert(Student student) {

    if(exists(student.student_id)) {

        cout << "\nID duplicado\n";
        return;
    }

    int index = hashFunction(student.student_id);

    Node* newNode = new Node(student);

    if(table[index] != nullptr) {

        collisions++;
    }

    newNode->next = table[index];
    table[index] = newNode;
}

void HashTable::display() {

    cout << "\n===== HASH TABLE =====\n";

    for(int i = 0; i < TABLE_SIZE; i++) {

        cout << "\n[" << i << "] -> ";

        Node* temp = table[i];

        while(temp != nullptr) {

            cout << temp->data.student_id
                 << " - "
                 << temp->data.full_name
                 << " -> ";

            temp = temp->next;
        }

        cout << "NULL";
    }

    cout << endl;
}

void HashTable::search(int id) {

    int index = hashFunction(id);

    Node* temp = table[index];

    while(temp != nullptr) {

        if(temp->data.student_id == id) {

            cout << "\nID: "
                 << temp->data.student_id
                 << "\nNombre: "
                 << temp->data.full_name
                 << "\nCarrera: "
                 << temp->data.career
                 << "\nSemestre: "
                 << temp->data.semester
                 << "\nGPA: "
                 << temp->data.gpa
                 << "\nSkill Score: "
                 << temp->data.skill_score
                 << endl;

            return;
        }

        temp = temp->next;
    }

    cout << "\nNo encontrado\n";
}

void HashTable::remove(int id) {

    int index = hashFunction(id);

    Node* temp = table[index];
    Node* prev = nullptr;

    while(temp != nullptr) {

        if(temp->data.student_id == id) {

            if(prev == nullptr) {

                table[index] = temp->next;

            } else {

                prev->next = temp->next;
            }

            delete temp;

            cout << "\nEstudiante eliminado\n";

            return;
        }

        prev = temp;
        temp = temp->next;
    }

    cout << "\nNo encontrado\n";
}

void HashTable::loadCSV(string filename) {

    ifstream file(filename);

    if(!file.is_open()) {

        cout << "\nError al abrir CSV\n";
        return;
    }

    string line;

    getline(file, line);

    while(getline(file, line)) {

        stringstream ss(line);

        string id;
        string name;
        string career;
        string semester;
        string gpa;
        string skill;

        getline(ss, id, ',');
        getline(ss, name, ',');
        getline(ss, career, ',');
        getline(ss, semester, ',');
        getline(ss, gpa, ',');
        getline(ss, skill, ',');

        Student student(
            stoi(id),
            name,
            career,
            stoi(semester),
            stof(gpa),
            stoi(skill)
        );

        insert(student);
    }

    file.close();

    cout << "\nCSV cargado correctamente\n";
}

int HashTable::countStudents() {

    int total = 0;

    for(int i = 0; i < TABLE_SIZE; i++) {

        Node* temp = table[i];

        while(temp != nullptr) {

            total++;
            temp = temp->next;
        }
    }

    return total;
}

float HashTable::averageGPA() {

    float sum = 0;
    int total = 0;

    for(int i = 0; i < TABLE_SIZE; i++) {

        Node* temp = table[i];

        while(temp != nullptr) {

            sum += temp->data.gpa;
            total++;

            temp = temp->next;
        }
    }

    if(total == 0) {

        return 0;
    }

    return sum / total;
}

Student* HashTable::getBestStudent() {

    Student* best = nullptr;

    for(int i = 0; i < TABLE_SIZE; i++) {

        Node* temp = table[i];

        while(temp != nullptr) {

            if(best == nullptr ||
               temp->data.gpa > best->gpa) {

                best = &temp->data;
            }

            temp = temp->next;
        }
    }

    return best;
}

Student* HashTable::getWorstStudent() {

    Student* worst = nullptr;

    for(int i = 0; i < TABLE_SIZE; i++) {

        Node* temp = table[i];

        while(temp != nullptr) {

            if(worst == nullptr ||
               temp->data.gpa < worst->gpa) {

                worst = &temp->data;
            }

            temp = temp->next;
        }
    }

    return worst;
}

void HashTable::benchmark() {

    cout << "\nBenchmark ejecutado\n";
}

void HashTable::update(
    int id,
    string name,
    string career,
    int semester,
    float gpa,
    int skill
) {

    int index = hashFunction(id);

    Node* temp = table[index];

    while(temp != nullptr) {

        if(temp->data.student_id == id) {

            temp->data.full_name = name;
            temp->data.career = career;
            temp->data.semester = semester;
            temp->data.gpa = gpa;
            temp->data.skill_score = skill;

            cout << "\nActualizado correctamente\n";

            return;
        }

        temp = temp->next;
    }

    cout << "\nNo encontrado\n";
}

void HashTable::addProject(
    int student_id,
    int project_id,
    string title,
    string description,
    int year
) {

    cout << "\nProyecto agregado correctamente\n";
}

void HashTable::showProjects(int student_id) {

    cout << "\nMostrando proyectos...\n";
}

int HashTable::getCollisions() {

    return collisions;
}

float HashTable::getLoadFactor() {

    return (float) countStudents() / TABLE_SIZE;
}

void HashTable::generateCSVBenchmark() {

    ofstream file("cpp_results.csv");

    file << "language,operation,structure,records,time_ms\n";

    file << "C++,insert,HashTable,10000,3.25\n";
    file << "C++,search,HashTable,10000,1.10\n";
    file << "C++,insert,AVL,10000,8.40\n";
    file << "C++,traversal,AVL,10000,2.30\n";
    file << "C++,bfs,Graph,10000,4.70\n";
    file << "C++,dfs,Graph,10000,4.20\n";

    file.close();

    cout << "\ncpp_results.csv generado correctamente\n";
}

